### ISSUE TITLE
Issues with Reproducing UnSAM Code

### ISSUE THREAD

**Issue by wenfang361** [frank-xwang/UnSAM#15]
Dear Author,

I have encountered some issues while working with your UnSAM code ([UnSAM/whole_image_segmentation/demo_whole_image.py at main · frank-xwang/UnSAM](https://github.com/frank-xwang/UnSAM)) and I would like to seek clarification on the following points:

1.Self-Supervised Training and Inference Code:

Since the GitHub page only provides a demo for inference, I initially proceeded with training on our dataset without paying attention to the configuration file UnSAM/whole_image_segmentation/configs/Base-COCO-InstanceSegmentation.yaml. This file contains a parameter SELF_TRAIN. Initially, I set this parameter to False and used a dataset that provided both images and annotations, resulting in supervised training. Later, I realized this parameter and changed it to True, and also updated DATASET_MAPPER_NAME to "sam_instance_tsv_self_train", corresponding to the dataset mapper file UnSAM/whole_image_segmentation/mask2former/data/dataset_mappers/sam_instance_tsv_self_train_dataset_mapper.py. However, after training with these changes, there were no results. I also noticed that the mask2former_model.py file does not seem to have a branch for self-supervised training, although the inference part does. I'm unsure if I have misunderstood something, but the self-supervised training failed, and I have significant doubts about the code. Is there a complete training script? Or am I missing something else?

2.Different Segmentation Tasks and Validation Issues:

While training, I noticed that by modifying the configuration file, I could change between different segmentation tasks. Initially, I followed the default setting for instance segmentation, but the actual goal is to perform semantic segmentation. When I changed the configuration to enable semantic segmentation and disable instance segmentation, the training still resulted in issues. The validation part yielded no results, and inference with the trained weights produced no usable masks. Additionally, I found that the validation script UnSAM/whole_image_segmentation/mask2former/evaluation/coco_evaluation.py can only evaluate instance segmentation and returns AP and AR results. However, for semantic segmentation, we want to compute the IOU value. Is there a branch or version of the code that supports this?

Also, regarding the num_classes parameter, my dataset contains three categories, including a background class. Should I exclude the background class when setting this parameter? I have set it to 2, is this correct?

I look forward to your guidance and assistance.

Best regards,

Jasmin

**Comment by frank-xwang** [comment]
Hi,  

Thank you for your message! I appreciate your questions, though I’m a bit confused about some of your questions. I’ll do my best to address them based on my current understanding—please feel free to let me know if anything is unclear or needs further clarification.  

---

### **1. Self-Supervised Training and Inference Code:**  
For training the model on downstream tasks, I don’t think any changes to the **.yaml** file are necessary. The **self_train** configuration is only used for model inference to generate predictions that can serve as pseudo ground-truths for the next round of training.  

If you already have annotations and images prepared, make sure the following two paths are correctly set:  
```
export DETECTRON2_DATASETS=/PATH/TO/YOUR/DATASETS  
export TRAIN_DATASETS=/PATH/TO/YOUR/SA-1B  
```  

Our implementation builds on **Mask2Former**:  
[https://github.com/facebookresearch/Mask2Former](https://github.com/facebookresearch/Mask2Former)  

You may find it helpful to review their dataset setup instructions to adapt it for your datasets:  
[Mask2Former Dataset Setup](https://github.com/facebookresearch/Mask2Former/tree/main/datasets)

---

### **2. Segmentation Tasks and Validation Issues:**  
I don’t recommend using other segmentation task settings with our codebase, as **UnSAM** is specifically designed for **unsupervised image segmentation**—both whole-image and promptable segmentation. It provides **hierarchical segmentation results**, ranging from subparts to parts, instances, and even overlapping instances. This is distinct from **semantic segmentation** tasks. 

If you plan to work with **semantic segmentation**, I suggest referring to the **[Getting Started Section](https://github.com/facebookresearch/Mask2Former/blob/main/GETTING_STARTED.md)** in the Mask2Former repository.  

If you have **human-annotated semantic segmentation masks**, you can follow the Mask2Former codebase to set up the training pipeline specifically for semantic segmentation tasks.  

---

### **3. Number of Classes:**  
The **number of classes should be set to 1**. For **unsupervised segmentation**, all masks are **class-agnostic**, similar to **SAM** and **SAM-2**.  

---

I hope this helps clarify your questions! Let me know if you need any additional assistance.

**Comment by wenfang361** [comment]
> Hi,
> 
> Thank you for your message! I appreciate your questions, though I’m a bit confused about some of your questions. I’ll do my best to address them based on my current understanding—please feel free to let me know if anything is unclear or needs further clarification.
> 
> ### **1. Self-Supervised Training and Inference Code:**
> For training the model on downstream tasks, I don’t think any changes to the **.yaml** file are necessary. The **self_train** configuration is only used for model inference to generate predictions that can serve as pseudo ground-truths for the next round of training.
> 
> If you already have annotations and images prepared, make sure the following two paths are correctly set:
> 
> ```
> export DETECTRON2_DATASETS=/PATH/TO/YOUR/DATASETS  
> export TRAIN_DATASETS=/PATH/TO/YOUR/SA-1B  
> ```
> 
> Our implementation builds on **Mask2Former**: https://github.com/facebookresearch/Mask2Former
> 
> You may find it helpful to review their dataset setup instructions to adapt it for your datasets: [Mask2Former Dataset Setup](https://github.com/facebookresearch/Mask2Former/tree/main/datasets)
> 
> ### **2. Segmentation Tasks and Validation Issues:**
> I don’t recommend using other segmentation task settings with our codebase, as **UnSAM** is specifically designed for **unsupervised image segmentation**—both whole-image and promptable segmentation. It provides **hierarchical segmentation results**, ranging from subparts to parts, instances, and even overlapping instances. This is distinct from **semantic segmentation** tasks.
> 
> If you plan to work with **semantic segmentation**, I suggest referring to the **[Getting Started Section](https://github.com/facebookresearch/Mask2Former/blob/main/GETTING_STARTED.md)** in the Mask2Former repository.
> 
> If you have **human-annotated semantic segmentation masks**, you can follow the Mask2Former codebase to set up the training pipeline specifically for semantic segmentation tasks.
> 
> ### **3. Number of Classes:**
> The **number of classes should be set to 1**. For **unsupervised segmentation**, all masks are **class-agnostic**, similar to **SAM** and **SAM-2**.
> 
> I hope this helps clarify your questions! Let me know if you need any additional assistance.

Thank you so much for taking the time to respond to my questions—I truly appreciate it. Your answer has been extremely helpful and has cleared up a lot of my confusion. Initially, I used your UnSAM work hoping to achieve a self-supervised semantic segmentation task, since I observed very promising results. However, I realized UnSAM seems more suitable for instance segmentation rather than semantic segmentation. I noticed there was a semantic segmentation branch in the code, so I assumed it could perform semantic segmentation. If I follow the training process provided in the Mask2Former repository for semantic segmentation, would that be considered fully supervised semantic segmentation? If so, that would differ from my original plan for a self-supervised approach, correct?

**Comment by frank-xwang** [comment]
If you want to do unsupervised semantic segmentation, I may recommend you to use my [U2Seg](https://github.com/u2seg/U2Seg) code, which is for Unsupervised Universal Image Segmentation. 

TL;DR: We present U2Seg, a unified framework for Unsupervised Universal image Segmentation that consistently outperforms previous state-of-the-art methods designed for individual tasks: CutLER for unsupervised instance segmentation, STEGO for unsupervised semantic segmentation, and the naive combination of CutLER and STEGO for unsupervised panoptic segmentation.

**Comment by Rudder573** [comment]
Dear Author,
       How prepare annotations and images for self-supervised training, are they the same as supervised ones, or do annotations make use of demo_dico.py-generated pseudo_masks_output.png?
